File: src/Lex.Web/MatchLanes.cs
using System.Text.Json.Nodes;
using static Lex.Web.PageShell;

namespace Lex.Web;

/// <summary>
/// The B2 lane classifier (agreed design plus the Codex ruling of 2026-08-28). A hit's served
/// match reasons place it in exactly one lane: text (provision evidence), identity (a name or
/// identifier answers "which law is this"), metadata (a subject association, never an answer)
/// or unclassified_render (an unknown reason renders through the existing visible path and is
/// never suppressed, but the code must not assert identity it has not classified). The
/// normative case table in tests/Lex.Tests/match-lane-cases.json binds this classifier and its
/// TypeScript twin to the same behavior.
/// </summary>
public static class MatchLanes
{
    public const string Text = "text";
    public const string Identity = "identity";
    public const string Metadata = "metadata";
    public const string UnclassifiedRender = "unclassified_render";

    // semantic_work and semantic_concept are deliberately NOT text: the producer's work
    // vector is subjects plus names, never provision text, and the concept arm is unreachable
    // and kind-unbound (Codex B2 review O3). Both fall through to unclassified_render.
    private static readonly HashSet<string> TextReasons =
        ["keyword", "semantic", "article_intent", "fuzzy"];

    private static readonly string[] IdentitySuffixes =
        ["_identifier", "_publisher_short_title", "_title", "_alias"];

    private const string AmbiguousPrefix = "ambiguous_";
    private const string MetadataReason = "work_metadata";

    public static string Classify(IReadOnlyList<string?> reasons)
    {
        if (reasons.Count == 0) return UnclassifiedRender;
        var sawIdentity = false;
        var sawMetadata = false;
        foreach (var raw in reasons)
        {
            var reason = raw ?? "";
            if (TextReasons.Contains(reason)) return Text;
            if (reason.StartsWith(AmbiguousPrefix, StringComparison.Ordinal))
            {
                // Ambiguity arises only during identity resolution: render, never suppress.
                sawIdentity = true;
                continue;
            }
            if (IdentitySuffixes.Any(suffix =>
                    reason.EndsWith(suffix, StringComparison.Ordinal)))
            {
                sawIdentity = true;
                continue;
            }
            if (string.Equals(reason, MetadataReason, StringComparison.Ordinal))
            {
                sawMetadata = true;
                continue;
            }
            // An unknown reason forbids every positive claim for this hit.
            return UnclassifiedRender;
        }
        if (sawIdentity) return Identity;
        return sawMetadata ? Metadata : UnclassifiedRender;
    }

    /// <summary>
    /// The response-level state: metadata_only holds only when at least one hit exists and
    /// every hit is POSITIVELY metadata (Codex Q1 ruling: unclassified never triggers the
    /// notice, it renders through the normal path).
    /// </summary>
    public static bool MetadataOnly(IReadOnlyList<IReadOnlyList<string?>> hitReasons) =>
        hitReasons.Count > 0
        && hitReasons.All(reasons => Classify(reasons) == Metadata);

    /// <summary>Decision 41 frozen copy for the metadata_only notice.</summary>
    public const string Heading = "No held text match";
    public const string Body =
        "Lex found records that match only in metadata. They are not shown as text answers. "
        + "This is not evidence that the named instrument or law does not exist. Check the name "
        + "or identifier, review coverage and known gaps, or search the official publisher.";
    public const string DisclosureLabel = "Matched only in metadata";

    /// <summary>A candidate disclosure row straight off served hit fields, validated here.</summary>
    public sealed record DisclosureRow(string Publisher, string Work, string ValidFrom, string Title);

    private static readonly System.Text.RegularExpressions.Regex WorkGrammar = new(
        "^[a-z0-9][a-z0-9._-]{0,199}$",
        System.Text.RegularExpressions.RegexOptions.IgnoreCase
        | System.Text.RegularExpressions.RegexOptions.CultureInvariant);

    private static readonly System.Text.RegularExpressions.Regex PublisherGrammar = new(
        "^[a-z0-9][a-z0-9-]{0,63}$",
        System.Text.RegularExpressions.RegexOptions.IgnoreCase
        | System.Text.RegularExpressions.RegexOptions.CultureInvariant);

    private static readonly System.Text.RegularExpressions.Regex DateGrammar = new(
        "^[0-9]{4}-[0-9]{2}-[0-9]{2}$",
        System.Text.RegularExpressions.RegexOptions.CultureInvariant);

    /// <summary>
    /// Every hit of a multi-publisher search response paired with its envelope publisher.
    /// Refused envelopes carry no hits array and simply contribute nothing, so a refusal
    /// beside metadata hits neither blocks nor fakes the response-level state.
    /// </summary>
    /// <summary>
    /// The only status under which a search envelope actually executed. Verified against the
    /// producer: the search case emits ok, retrieval_mode_unavailable, unknown_work,
    /// unknown_anchor and no_provision_history, and never no_result. Round 1 admitted
    /// no_result, so a cross-operation envelope carrying metadata hits could authorize
    /// suppression (B1+B2 round 2 review, O1).
    /// </summary>
    private static readonly HashSet<string> SearchSuccessStatuses = ["ok"];

    /// <summary>
    /// The authoritative population plus whether it is COMPLETE. Round 1 discarded response
    /// invalidity: a successful envelope whose hits field was malformed was read as empty, so
    /// a sibling metadata-only response could still authorize suppression (B1+B2 round 2
    /// review, O2). An incomplete population makes the positive claim unreachable.
    /// </summary>
    public static (IReadOnlyList<(string Publisher, JsonObject Hit)> Rows, bool Complete)
        ResponsePopulation(JsonArray envelopes)
    {
        var rows = new List<(string, JsonObject)>();
        var complete = true;
        foreach (var node in envelopes)
        {
            if (node is not JsonObject result) { complete = false; continue; }
            var envelope = result["envelope"] as JsonObject;
            var status = envelope?["status"] is JsonValue statusValue
                && statusValue.TryGetValue<string>(out var statusText) ? statusText : null;
            if (status is null) { complete = false; continue; }
            // Only an authoritative successful envelope may contribute to the positive
            // metadata_only claim; a refusal's rows are not evidence.
            if (!SearchSuccessStatuses.Contains(status)) continue;
            if (result["hits"] is not JsonArray hits) { complete = false; continue; }
            var publisher = envelope!["publisher"] is JsonValue value
                && value.TryGetValue<string>(out var text) ? text : "";
            foreach (var hitNode in hits)
            {
                if (hitNode is not JsonObject hit) { complete = false; continue; }
                // A reasons field that is present but not an array of strings is malformed
                // evidence, not absent evidence.
                if (hit["match_reasons"] is { } reasonsNode
                    && (reasonsNode is not JsonArray reasons
                        || reasons.Any(reason => reason is not JsonValue member
                            || !member.TryGetValue<string>(out _))))
                {
                    complete = false;
                    continue;
                }
                rows.Add((publisher, hit));
            }
        }
        return (rows, complete);
    }

    /// <summary>
    /// True when any envelope reports a truncated row set (B1+B2 review, O4). The producer
    /// carries response_row_set.truncated; an exact overflow count would be a claim the
    /// response cannot support, so the countersigned fallback sentence is used instead.
    /// </summary>
    public static bool AnyRowSetTruncated(JsonArray envelopes) =>
        envelopes.OfType<JsonObject>().Any(result =>
            (result["response_row_set"] as JsonObject)?["truncated"] is JsonValue value
            && value.TryGetValue<bool>(out var truncated) && truncated);

    /// <summary>
    /// The server-page notice plus the subordinate disclosure list, for the ONE response-level
    /// metadata_only state (Codex B2 review, O1: classification is never per publisher).
    /// Rows are validated fail closed against the B1 coordinate grammar and links are rebuilt
    /// from the parsed parts; invalid rows are omitted without suppressing the notice (O4).
    /// One exact-host official action renders per represented collection. Byte-exact
    /// boundaries and append-only insertion, per the B1 classifier finding.
    /// </summary>
    public static string NoticeHtml(
        IReadOnlyList<string> collections,
        IReadOnlyList<DisclosureRow> rows,
        bool truncated = false)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var valid = rows.Where(row =>
                PublisherGrammar.IsMatch(row.Publisher)
                && WorkGrammar.IsMatch(row.Work)
                && DateGrammar.IsMatch(row.ValidFrom)
                && seen.Add($"{row.Publisher}:{row.Work}"))
            .ToArray();
        var items = string.Join("", valid.Take(10).Select(row =>
        {
            var title = row.Title.Length is > 0 and <= 300 ? row.Title
                : row.Title.Length > 300 ? row.Title[..300] : row.Work;
            return $"<li><a href=\"/{H(row.Publisher)}/{H(row.Work)}/{H(row.ValidFrom)}\">"
                + $"{H(title)}</a> <span class=\"sub mono\">{H(row.Work)} · {H(row.Publisher)}"
                + " · matched in metadata</span></li>";
        }));
        // C3 ruling: N counts only valid, logically deduplicated suppressed matches present
        // in this bounded response, minus the rows shown; never a corpus-wide claim. Search
        // envelopes carry no truncation marker today, so the exact count always exists.
        // C3: an exact N is only honest when the response is complete. A truncated row set
        // holds no exact total, so the countersigned fallback sentence is used instead of
        // inventing one.
        var overflow = truncated
            ? "<span class=\"sub\">additional returned matches are not shown</span>"
            : valid.Length <= 10 ? ""
            : $"<span class=\"sub\">and {valid.Length - 10} more returned matches</span>";
        var disclosure = valid.Length == 0 ? "" :
            $"<details><summary>{H(DisclosureLabel)}</summary><ul>{items}</ul>{overflow}</details>";
        var officials = string.Join("", collections
            .Where(collection => PublisherGrammar.IsMatch(collection))
            .Distinct(StringComparer.Ordinal)
            .OrderBy(collection => collection, StringComparer.Ordinal)
            .Select(WorkCandidates.OfficialSearchHref)
            .Distinct(StringComparer.Ordinal)
            .Select(official => official.StartsWith("https://", StringComparison.Ordinal)
                ? $" &nbsp;&nbsp;<a href=\"{H(official)}\" rel=\"noopener\">Search the official publisher</a>"
                : $" &nbsp;&nbsp;<a href=\"{H(official)}\">Search Lex</a>"));
        return "<div class=\"notice\" role=\"note\" data-testid=\"metadata-only-notice\""
            + $" aria-label=\"{H(Heading)}\">"
            + $"<b>{H(Heading)}.</b> {H(Body)}"
            + disclosure
            + $"<span class=\"sub\"><a href=\"/coverage\">View coverage and known gaps</a>"
            + $"{officials}</span>"
            + "</div>";
    }

    /// <summary>
    /// Served reasons for one hit, read fail-closed (B1+B2 review, O5). A missing array is
    /// unclassified. A member that is not a JSON string is mapped to null rather than read
    /// through GetValue&lt;string&gt;, which throws on a number or object and would take the
    /// whole page down; null is an unknown reason, so the hit renders and is never suppressed.
    /// </summary>
    public static IReadOnlyList<string?> ReasonsOf(JsonObject hit) =>
        hit["match_reasons"] is JsonArray reasons
            ? reasons.Select(reason =>
                reason is JsonValue value && value.TryGetValue<string>(out var text)
                    ? text : null).ToArray()
            : [];
}

File: web/src/matchLanes.ts
/**
 * The B2 lane classifier, TypeScript twin of src/Lex.Web/MatchLanes.cs. Both are bound to the
 * one normative case table (tests/Lex.Tests/match-lane-cases.json); the parity test loads that
 * exact file. A hit is text, identity, metadata, or unclassified_render; an unknown reason
 * renders through the existing visible path and is never suppressed, and never asserted as
 * identity (Codex Q1 ruling, 2026-08-28).
 */

export type MatchLane = "text" | "identity" | "metadata" | "unclassified_render";

// semantic_work and semantic_concept are deliberately NOT text: the producer's work vector
// is subjects plus names, never provision text, and the concept arm is unreachable and
// kind-unbound (Codex B2 review O3). Both fall through to unclassified_render.
const TEXT_REASONS = new Set(["keyword", "semantic", "article_intent", "fuzzy"]);
const IDENTITY_SUFFIXES = ["_identifier", "_publisher_short_title", "_title", "_alias"];
const AMBIGUOUS_PREFIX = "ambiguous_";
const METADATA_REASON = "work_metadata";

export function classifyMatchLane(reasons: readonly (string | undefined | null)[]): MatchLane {
  if (reasons.length === 0) return "unclassified_render";
  let sawIdentity = false;
  let sawMetadata = false;
  for (const raw of reasons) {
    const reason = raw ?? "";
    if (TEXT_REASONS.has(reason)) return "text";
    if (reason.startsWith(AMBIGUOUS_PREFIX)) { sawIdentity = true; continue; }
    if (IDENTITY_SUFFIXES.some((suffix) => reason.endsWith(suffix))) {
      sawIdentity = true;
      continue;
    }
    if (reason === METADATA_REASON) { sawMetadata = true; continue; }
    return "unclassified_render";
  }
  if (sawIdentity) return "identity";
  return sawMetadata ? "metadata" : "unclassified_render";
}

/** metadata_only holds only when every hit is POSITIVELY metadata; unclassified never triggers it. */
export function metadataOnlyState(
  hitReasons: readonly (readonly (string | undefined | null)[])[],
): boolean {
  return hitReasons.length > 0
    && hitReasons.every((reasons) => classifyMatchLane(reasons) === "metadata");
}

/**
 * A served reason list straight off the wire: a non-array or any non-string member makes the
 * hit unclassified_render, never metadata, so a hostile or wrong-typed shape can neither
 * crash rendering nor suppress results (Codex B2 review, O4).
 */
export function laneOfServedReasons(raw: unknown): MatchLane {
  if (!Array.isArray(raw) || raw.some((member) => typeof member !== "string")) {
    return "unclassified_render";
  }
  return classifyMatchLane(raw as string[]);
}

/**
 * The response-level decision, taken over the COMPLETE fused population before any display
 * cap, dedup projection or passage filter (Codex B2 review, O2). Hits that collapse to one
 * logical work contribute the UNION of their reasons, so a metadata row can never mask a
 * text or identity row for the same work. Any wrong-typed reason shape forbids the
 * metadata_only claim outright.
 */
export function metadataOnlyResponse(
  population: readonly { work: unknown; reasons: unknown }[],
): boolean {
  if (population.length === 0) return false;
  const unionByWork = new Map<string, string[]>();
  for (const [index, hit] of population.entries()) {
    if (!Array.isArray(hit.reasons)
      || hit.reasons.some((member) => typeof member !== "string")) {
      return false;
    }
    const key = typeof hit.work === "string" && hit.work.length > 0
      ? hit.work
      : `#anonymous-${index}`;
    unionByWork.set(key, [...(unionByWork.get(key) ?? []), ...(hit.reasons as string[])]);
  }
  return [...unionByWork.values()]
    .every((reasons) => classifyMatchLane(reasons) === "metadata");
}

/** One logical work in the authoritative response population. */
export interface PopulationEntry {
  work: string;
  title: string;
  reasons: unknown;
}

/** Statuses under which a search envelope actually executed and may contribute evidence. */
const SEARCH_SUCCESS_STATUSES = new Set(["ok", "no_result"]);

/**
 * The authoritative response population, read from the RAW envelopes (B1+B2 review, O2 and
 * O5). Three things must all be true here and cannot be recovered later:
 *
 * 1. Only an envelope with a successful status contributes. A refused or malformed envelope's
 *    rows are not evidence, and admitting them can suppress real answers behind the notice.
 * 2. Reasons are unioned per logical work BEFORE any fusion. The workspace's fusion step
 *    deduplicates by identity and discards the losing hit's match_reasons, so a work matched
 *    by metadata in one publisher and by keyword in another would otherwise read as
 *    metadata-only and be suppressed.
 * 3. The full population is retained, not the display slice, so the disclosure and its
 *    overflow count describe the whole response.
 */
export function responsePopulation(raw: unknown): PopulationEntry[] {
  const envelopes = Array.isArray(raw) ? raw : [raw];
  const byWork = new Map<string, PopulationEntry>();
  for (const envelope of envelopes) {
    if (typeof envelope !== "object" || envelope === null) continue;
    const record = envelope as Record<string, unknown>;
    const meta = record.envelope as Record<string, unknown> | undefined;
    const status = typeof meta?.status === "string" ? meta.status : undefined;
    if (status === undefined || !SEARCH_SUCCESS_STATUSES.has(status)) continue;
    const hits = Array.isArray(record.hits) ? record.hits : [];
    for (const hit of hits) {
      if (typeof hit !== "object" || hit === null) continue;
      const row = hit as Record<string, unknown>;
      const work = String(row.lex_id ?? "").split(":").slice(0, 2).join(":");
      if (!work) continue;
      const reasons = Array.isArray(row.match_reasons) ? row.match_reasons : [];
      const existing = byWork.get(work);
      if (existing) {
        // Union, never replace: a later hit's reasons are as authoritative as the first's.
        existing.reasons = [
          ...(existing.reasons as unknown[]),
          ...reasons,
        ];
        if (!existing.title && typeof row.title === "string") existing.title = row.title;
        continue;
      }
      byWork.set(work, {
        work,
        title: typeof row.title === "string" ? row.title : work,
        reasons: [...reasons],
      });
    }
  }
  return [...byWork.values()];
}

/** True when any envelope reports a truncated row set, so no exact overflow total exists. */
export function anyRowSetTruncated(raw: unknown): boolean {
  const envelopes = Array.isArray(raw) ? raw : [raw];
  return envelopes.some((envelope) => {
    if (typeof envelope !== "object" || envelope === null) return false;
    const rowSet = (envelope as Record<string, unknown>).response_row_set;
    return typeof rowSet === "object" && rowSet !== null
      && (rowSet as Record<string, unknown>).truncated === true;
  });
}

/**
 * The official publisher search entry per collection, exact reviewed hosts only; an unknown
 * collection falls back to the internal search page rather than guessing a URL. Mirrors the
 * server-side authority in WorkCandidates.OfficialSearchHref.
 */
export function officialSearchHref(collection: string): string {
  switch (collection) {
    case "lu-legilux": return "https://legilux.public.lu";
    case "eu-eurlex": return "https://eur-lex.europa.eu";
    default: return "/search";
  }
}

/** Decision 41 frozen copy, browser authority for the workspace surface. */
export const METADATA_ONLY_HEADING = "No held text match";
export const METADATA_ONLY_BODY =
  "Lex found records that match only in metadata. They are not shown as text answers. This "
  + "is not evidence that the named instrument or law does not exist. Check the name or "
  + "identifier, review coverage and known gaps, or search the official publisher.";
export const METADATA_ONLY_DISCLOSURE = "Matched only in metadata";
